File: 3-longest-substring-without-repeating-characters/longest-substring-without-repeating-characters.cpp
class Solution {
public:
    int lengthOfLongestSubstring(string s) {
        int n=s.size();
        vector<int> hash(256,-1);
        int r=0,l=0,maxlen=0;
        while(r<n){
            if(hash[(int)s[r]]!=-1){
                if(hash[(int)s[r]]>=l){
                    l=hash[(int)s[r]]+1;
                }
            }
            int len=r-l+1;
            maxlen=max(maxlen,len);
            hash[(int)s[r]]=r;
            r++;
        }
        return maxlen;
        
    }
};

File: 3-longest-substring-without-repeating-characters/longest-substring-without-repeating-characters.java
class Solution {
    public int lengthOfLongestSubstring(String s) {
        int n = s.length();
        int[] hash = new int[256];

        // Initialize all values to -1
        for (int i = 0; i < 256; i++) {
            hash[i] = -1;
        }

        int left = 0;
        int maxLen = 0;
        int right=0;
        if(n==0) return 0;
        while(left<=right){
            char ch = s.charAt(right);

            if (hash[ch] != -1 && hash[ch] >= left) {
                left = hash[ch] + 1;
            }

            int len = right - left + 1;
            maxLen = Math.max(maxLen, len);

            hash[ch] = right;
            if(right<n-1){
                right++;
            }
            
        }

        return maxLen;
    }
}